### PR TITLE
Make Object.assign() safe to use with simple partial sources

### PR DESCRIPTION
**Pull request by gaearon**
Fixes https://github.com/facebook/prepack/issues/1462.

Supersedes my attempts in https://github.com/facebook/prepack/pull/1459, https://github.com/facebook/prepack/pull/1460, and https://github.com/facebook/prepack/pull/1468.
The git history reflects a lot of back-and-forth so it might make more sense to look at the final code.

There are two separate fixes here:

* If we have a partial simple source and then a normal source, we shouldn't attempt to `Set` any further keys after the first partial. Instead we should emit residual `Object.assign` calls. If we set the keys, we end up with a wrong key order (a key gets set earlier than it should have). Key order (AFAIK) isn't mandated by the spec but tends to be relied on in product code because all engines preserve it in the order of assignments.

* It is not safe to mutate either a partial source or the target (when some sources are partial) in Prepack land after the `Object.assign` call. This is because Prepack will serialize all those mutations (as the object final state) right into the residual `Object.assign` call. But this is wrong because they haven't happened yet.

The second issue is thorny. In many cases sources (and the target) don't get mutated later. So we'd like to keep supporting `Object.assign` with partials when we don't touch them later.

To solve it, I introduced new `obj.makeFinal()` and `obj.isFinalObject()` methods. If an object is marked as "final", we cannot make changes to it that would change its emitted form. This is a way for us to say "let's try to continue as far as we can, but if anything tries to touch it, fatal".

It doesn't mean the object is "frozen" (technically we didn't freeze it), but it means Prepack can't deal with further mutations to it because Prepack has already "used" its current snapshot and it would be unsafe to change it.

I'm not sure I'm checking `isFinalObject()` in all places where it's necessary. If there is a way to do this with stronger guarantees please let me know. I want to still highlight that this fixes multiple existing issues on master so it's a step in a safer duration, even though I'm wary that if the concept is not fully baked, it might give a false sense of security.

Here's how we're using this new infra. When we emit a residual `Object.assign()` call because some sources are partial (and simple), I mark all sources and the target as "final". So if they don't get mutated, everything is fine. If they get mutated later, we get a fatal when they do. In the future Prepack might support this better, but the fatal makes it clear which cases it currently doesn't handle (and would produce wrong output for).

There might also be other cases where making an object as "final" lets us continue further, although I don't know Prepack codebase well enough to tell for sure. So I'm hoping it is useful as a generic concept outside of the `Object.assign()` use case.

The pure mode is treated a bit differently. In the pure mode we can get away with "leaking" final values whenever we need to write to them. Then they're treated like any other "leaked" object.

I added regression tests for all existing cases I could find, both in pure and impure mode.

### CODE REVIEW

File: src/values/ObjectValue.js [hunk 1]
```diff
@@ -244,6 +246,10 @@ export default class ObjectValue extends ConcreteValue {
   // to return AbstractValue for unknown properties.
   _isSimple: BooleanValue;
 
+  // If true, it is not safe to perform any more mutations that would change
+  // the object's serialized form.
+  _isFinal: BooleanValue;
```
Comment (NTillmann) [inline src/values/ObjectValue.js]: All of the "tracked" properties should include `| AbstractValue` in their type, as merging two different states of the flag might result in an `AbstractValue`. (There's an open issue for that: #1124.)
In that sense, the current handling of `_isPartial` and `_isSimple` is fragile, but at least so far making objects partial or simple is a meta statement that needs to be explicitly inserted by some framework, but your setting of `_isFinal` may happen implicitly as part of the regular program execution flow.

Comment (gaearon) [reply src/values/ObjectValue.js]: Thanks for explaining. I wasn't sure what "tracking" means but this makes more sense now.

Comment (gaearon) [reply src/values/ObjectValue.js]: Does this mean the whole approach is flawed, or just that I should make it accept `AbstractValue` and also treat this as a bailout reason when I check for it?

Comment (NTillmann) [reply src/values/ObjectValue.js]: The approach should be fine, it just needs to handle AbstractValues and bail out when it runs into issues.

**Comment by sebmarkbage** [comment]
For context, I suggested not doing down the snapshot strategy because that's what I originally tried in #1142 and that turned out to be very complicated to get in properly in the current model. It was easy to get it wrong (and it was the wrong strategy for leaking anyway). So I suggested banning further modifications to unblock us for now since this covers our current needs. Even though the long term strategy could be to model snapshots.

File: src/intrinsics/ecma262/Object.js [hunk 2]
```diff
@@ -152,6 +170,13 @@ export default function(realm: Realm): NativeFunctionValue {
       // We already established above that `to` is simple,
       // so set the `_isSimple` flag.
       to.makeSimple();
+
+      if (to instanceof ObjectValue) {
```
[the remark this answers is not in the document]
Comment (gaearon) [reply src/intrinsics/ecma262/Object.js]: In which cases does this matter? Everywhere I check for `isFinalObject`, the argument is already just `ObjectValue`. Am I missing some checks?

**Comment by gaearon** [comment]
@hermanventer @NTillmann 

Mind another review? I believe I addressed the comments in the last two commits.

**Comment by gaearon** [comment]
Pushed up a Flow fix.

Regarding the error message, `realm.currentLocation` seems sufficient. Is it not? In all error handler tests I added, the error points to the exact line with the mutation.

I did add a separate error code though.

File: src/methods/properties.js [hunk 3]
```diff
@@ -211,21 +211,25 @@ function ensureIsNotFinal(realm: Realm, O: ObjectValue, P: void | PropertyKeyVal
...
+    if (O.isLeakedObject()) {
+      return;
+    }
   }
-  // The object was created outside of pure scope so we couldn't leak.
-  // Give up.
-  AbstractValue.reportIntrospectionError(O, P);
+  // We can't continue because this object is already in its final state.
+  let error = new CompilerDiagnostic(
+    "Mutating an object that has been marked as final is not yet supported " +
```
Comment (hermanventer) [inline src/methods/properties.js]: "Marked as final" is an implementation detail that is not part of the mental model of someone using Prepack. Consider a message like "Mutating an object with unknown properties, after some of those properties have already been used, is not supported." In the wiki page you can can discuss Object.assign and perhaps give a code snippet illustrating how things can go wrong.

Comment (gaearon) [reply src/methods/properties.js]: Addressed

**Comment by gaearon** [comment]
Filed https://github.com/facebook/prepack/issues/1480 as a followup.

**Pull request closed without merging**